Add randomised Quickcheck testing for StackValue instances

ajnsit opened the pull request:
Added some Quickcheck randomised tests for testing StackValue instances (as mentioned in https://github.com/osa1/hslua/issues/31)

osa1 commented on test/HsLuaSpec.hs:
```diff
@@ -69,3 +83,55 @@ nulString =
+             x2 <- peek l (-4)
+             return (x1, x2)
+  QM.assert $ xs == (fromJust x1)
+  QM.assert $ xs == (fromJust x2)
```
If we introduce a bug in the future, `fromJust` will fail with horrible error message because it's a partial function that we expect to work fine here, assuming that we don't have a bug. Instead, wrapping `xs` with `Just` would be better here, because in that case we'd get a nice error message if we introduce a bug.

i.e.

```haskell
QM.assert $ Just xs == fromJust x1
```

Same for the next line.

osa1 commented on test/HsLuaSpec.hs:
```diff
@@ -9,6 +10,12 @@ import Test.Hspec
+import qualified Test.QuickCheck.Monadic as QM
+import Test.QuickCheck.Instances ()
+
+import Data.Maybe (fromJust)
```
After the changes I below we can get rid of this line.

osa1 commented on test/HsLuaSpec.hs:
```diff
@@ -69,3 +83,55 @@ nulString =
+         l <- newstate
+         push l xs
+         peek l (-1)
+  QM.assert $ xs == (fromJust x)
```
Similarly, get rid of `fromJust`.

osa1 commented:
Thanks for doing this.

Could you also add some assertions for checking the stack size after push operations? We're occasionally having bugs about this. (see recent list pop method bug fixed with #30)

ajnsit commented:
I rewrote the tests to test pushing and peeking at random indices, instead of fixed at 1 and 4. I also added checks for stack size.

However, that seems to have uncovered a C bug. 

Now the tests fail with a malloc error when pushing bytestrings. It looks like this on my machine -

    HsLua
      StackValue
        ByteString -- unicode stuff
        ByteString should survive after GC/Lua destroyed
        Push/pop StackValue lists
        String with NUL byte should be pushed/popped correctly
      Random StackValues
        can push/pop booleans
        can push/pop ints
    *** Error in `/home/aj/consulting/foss/hslua/.stack-work/dist/x86_64-linux/Cabal-1.18.1.5/build/test/test': malloc(): smallbin double linked list corrupted: 0x00007f5168064290 ***
    Test suite failure for package hslua-0.4.1
        test:  exited with: ExitFailure (-6)
    Logs printed to console

Because it fails in C code, we don't get useful logs. However, some debugging tells me that the test fails at values like this -

    Testing value ------------------------------- "\233F(\253"
    Inserted at indices: [2,4,4,5,5,8,8,13,13,16,17,19,19,21]
    Total StackSize: 21

I have left in the debugging statements in the code so you can simply uncomment them to check this on your machine.

Also, it seems like the test programs are also failing with logs like the following -

    Linking .stack-work/dist/x86_64-linux/Cabal-1.18.1.5/build/test/test ...
    hslua-0.4.1: test (suite: callbacks)
    begin
    callback 1
    callback 2
    reset callbacks
    add callbacks in reverse order
    callback 3
    callback 2
    callback 1
    Failing from Lua
    3
    false
    callback 1 return value
    [string "examples/callbacks/callbacks.lua"]:18: failed
    end
    hslua-0.4.1: test (suite: err_prop)
    Lua: 10
    Haskell: 9
    Lua: 8
    Haskell: 7
    Lua: 6
    Haskell: 5
    Lua: 4
    Haskell: 3
    Lua: 2
    Haskell: 1
    Lua: 0
    ret: 2
    errMsg: [string "examples/err_prop/err_prop.lua"]:4: [string "examples/err_prop/err_prop.lua"]:4: [string "examples/err_prop/err_prop.lua"]:4: [string "examples/err_prop/err_prop.lua"]:4: [string "examples/err_prop/err_prop.lua"]:4: [string "examples/err_prop/err_prop.lua"]:12: Failing from Lua
    top: 1
    Haskell: 10
    Lua: 9
    Haskell: 8
    Lua: 7
    Haskell: 6
    Lua: 5
    Haskell: 4
    Lua: 3
    Haskell: 2
    Lua: 1
    Haskell: 0
    errMsg: [string "examples/err_prop/err_prop.lua"]:4: [string "examples/err_prop/err_prop.lua"]:4: [string "examples/err_prop/err_prop.lua"]:4: [string "examples/err_prop/err_prop.lua"]:4: [string "examples/err_prop/err_prop.lua"]:4: Failing from Haskell
    hslua-0.4.1: test (suite: haskellfun)
    hello world!
    335.54432
    Hello, World!
    Error caught from Haskell land: argument 1 of Haskell function: number expected, got string

    Error caught from Haskell land: argument 2 of Haskell function: number expected, got no value

I have not investigated this further, so I don't know what the problem with the test programs is.

osa1 commented:
Wow, this is messed up! I'm running some experiments to understand the issue here and I'm getting stack traces like:

```
λ:19> main
*** Error in `/usr/local/lib64/ghc-7.10.1/bin/ghc': free(): invalid pointer: 0x00007f718c2aa3c0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x730bf)[0x7f71969350bf]
/lib64/libc.so.6(+0x7892e)[0x7f719693a92e]
/lib64/libc.so.6(+0x79636)[0x7f719693b636]
/usr/local/lib64/ghc-7.10.1/bin/../rts/libHSrts_thr-ghc7.10.1.so(+0x35572)[0x7f71970c9572]
/usr/local/lib64/ghc-7.10.1/bin/../rts/libHSrts_thr-ghc7.10.1.so(+0x3565c)[0x7f71970c965c]
/usr/local/lib64/ghc-7.10.1/bin/../rts/libHSrts_thr-ghc7.10.1.so(+0x3c2bd)[0x7f71970d02bd]
/usr/local/lib64/ghc-7.10.1/bin/../rts/libHSrts_thr-ghc7.10.1.so(+0x2d531)[0x7f71970c1531]
/usr/local/lib64/ghc-7.10.1/bin/../rts/libHSrts_thr-ghc7.10.1.so(+0x2e18c)[0x7f71970c218c]
/usr/local/lib64/ghc-7.10.1/bin/../rts/libHSrts_thr-ghc7.10.1.so(scheduleWaitThread+0x98)[0x7f71970c3248]
/usr/local/lib64/ghc-7.10.1/bin/../rts/libHSrts_thr-ghc7.10.1.so(hs_main+0x7b)[0x7f71970bfb1b]
/usr/local/lib64/ghc-7.10.1/bin/ghc[0x50c9fb]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7f71968e3b05]
/usr/local/lib64/ghc-7.10.1/bin/ghc[0x41cc39]
```

I'm wondering if this could be related with GHC RTS. I'll be debugging this.

osa1 commented:
Using non-threaded RTS helped with debugging:

```
Program received signal SIGSEGV, Segmentation fault.
0x00000000006382a8 in luaS_newlstr ()
Missing separate debuginfos, use: zypper install libgcc_s1-debuginfo-4.8.3+r212056-2.2.4.x86_64 libgmp10-debuginfo-5.1.3-3.1.2.x86_64
(gdb) bt
#0  0x00000000006382a8 in luaS_newlstr ()
#1  0x000000000063db2a in lua_pushlstring ()
#2  0x0000000000410421 in cf9A_info ()
#3  0x0000000000000000 in ?? ()
```

Let's see what this means..

osa1 commented:
Merged, thanks! You can see the fix here: https://github.com/osa1/hslua/commit/0dbe42655b9929de03b0327d608878744b33314a

ajnsit commented:
Great! Thanks!
